### PR TITLE
chore: refactor custom token

### PR DESCRIPTION
**Pull request by nguyennh4522**
# Solution

## Note
- Cannot add native token
- Cannot add tokens that are not suppoted by bot

## How to use

- Command: `$tokens add-custom <address> <symbol> <chain>` 
  - `address`: contract address
  - `symbol`: symbol of token
  - `chain`: currency of chain
- Example: `$tokens add-custom 0x0e09fabb73bd3ade0a17ecc321fd13a19e81ce82 cake bnb`

- Images:
image

### CODE REVIEW

File: pkg/entities/tokens.go [hunk 1]
```diff
@@ -2,92 +2,98 @@ package entities
...
 )
 
-type Response struct {
-	Id     string `json:"id"`
-	Symbol string `json:"symbol"`
-	Name   string `json:"name"`
-}
+func (e *Entity) CreateCustomToken(req request.UpsertCustomTokenConfigRequest) error {
+	token, _ := e.repo.Token.GetOneBySymbol(req.Symbol)
+	if token == nil {
```
Comment (trkhoi) [inline pkg/entities/tokens.go:17]: log error

File: pkg/entities/tokens.go [hunk 2]
```diff
@@ -2,92 +2,98 @@ package entities
...
-		return false, err
+	if !token.DiscordBotSupported {
+		return errors.New("token is not supported to add to your server")
 	}
 
-	for i, _ := range listSymbol {
-		if strings.ToLower(symbol) == strings.ToLower(listSymbol[i].Symbol) {
-			return true, nil
-		}
+	fmt.Println(token.Chain.Currency, req.Chain)
```
Comment (trkhoi) [inline pkg/entities/tokens.go:50]: remove this

File: pkg/entities/tokens.go [hunk 3]
```diff
@@ -2,92 +2,98 @@ package entities
...
-		Symbol:              strings.ToUpper(req.Symbol),
-		ChainID:             req.ChainID,
-		Decimals:            req.Decimals,
-		DiscordBotSupported: req.DiscordBotSupported,
-		CoinGeckoID:         req.CoinGeckoID,
-		Name:                strings.ToUpper(req.Name),
-		GuildDefault:        req.GuildDefault,
-	})
+func (e *Entity) GetAllSupportedToken(guildID string) (returnToken []model.Token, err error) {
+	returnToken, err = e.repo.Token.GetAllSupportedToken(guildID)
```
Comment (trkhoi) [inline pkg/entities/tokens.go:64]: log error everywhere

**Review by trkhoi — CHANGES_REQUESTED** [review]

File: pkg/entities/tokens.go (hunk 3 above)
Comment (nguyennh4522) [reply pkg/entities/tokens.go:64]: this is old code, I don't change anything here

File: pkg/entities/tokens.go (hunk 1 above)
Comment (nguyennh4522) [reply pkg/entities/tokens.go:21]: resolved

File: pkg/entities/tokens.go (hunk 2 above)
Comment (nguyennh4522) [reply pkg/entities/tokens.go]: resolved

File: pkg/entities/tokens.go [hunk 4]
```diff
@@ -91,3 +79,32 @@ func (e *Entity) GetAllSupportedToken(guildID string) (returnToken []model.Token
 
 	return returnToken, nil
 }
+
+func getCoinGeckoBySymbol(symbol string) (*response.CoinGekoResponse, error) {
```
Comment (anhnh12) [inline pkg/entities/tokens.go:83]: why get by symbol but using `/api/v3/coins/list`?

Comment (anhnh12) [inline pkg/entities/tokens.go:83]: move to `coingecko` service

File: pkg/entities/tokens.go [hunk 5]
```diff
@@ -2,85 +2,73 @@ package entities
...
-		if strings.ToLower(symbol) == strings.ToLower(listSymbol[i].Symbol) {
-			return true, nil
-		}
+	if !token.DiscordBotSupported {
+		return errors.New("token is not supported to add to your server")
 	}
 
-	return false, nil
-}
+	if !strings.EqualFold(token.Chain.Currency, req.Chain) {
```
Comment (anhnh12) [inline pkg/entities/tokens.go:54]: why compare chain currency with chain?

File: pkg/entities/tokens.go [hunk 6]
```diff
@@ -91,3 +79,32 @@ func (e *Entity) GetAllSupportedToken(guildID string) (returnToken []model.Token
 
 	return returnToken, nil
 }
+
+func getCoinGeckoBySymbol(symbol string) (*response.CoinGekoResponse, error) {
+	req, err := http.NewRequest(http.MethodGet, "https://api.coingecko.com/api/v3/coins/list", nil)
```
Comment (anhnh12) [inline pkg/entities/tokens.go:84]: try using `SearchCoin` in `coingecko.go` instead

File: pkg/entities/tokens.go (hunk 5 above)
Comment (nguyennh4522) [reply pkg/entities/tokens.go:54]: because req.Chain is currency value of chain

File: pkg/entities/tokens.go (hunk 4 above)
Comment (nguyennh4522) [reply pkg/entities/tokens.go]: removed

Comment (nguyennh4522) [reply pkg/entities/tokens.go]: resolved

File: pkg/entities/tokens.go (hunk 6 above)
Comment (nguyennh4522) [reply pkg/entities/tokens.go]: resolved

File: pkg/entities/tokens.go [hunk 7]
```diff
@@ -48,7 +48,7 @@ func (e *Entity) CreateCustomToken(req request.UpsertCustomTokenConfigRequest) e
 		return errors.New("token is not supported to add to your server")
 	}
 
-	if !strings.EqualFold(token.Chain.Currency, req.Chain) {
+	if !strings.EqualFold(token.Chain.ShortName, req.Chain) {
```
Comment (anhnh12) [inline pkg/entities/tokens.go:51]: find chain by `req.Chain` -> useless comparison?

**Review by anhnh12 — APPROVED** [review]

**Pull request merged**